checker/src/call_visitor.rs

```diff
@@ -1994,4 +2127,51 @@ impl<'call, 'block, 'analysis, 'compilation, 'tcx, E>
         }
         Rc::new("dummy argument".to_string())
     }
+
+    /// Extract the tag kind and the control mask from the generic arg of the function call
+    /// underlying `add_tag!` or `has_tag!`. The tag type should be the second generic argument
+    /// of the current function call. The tag type itself should also be parametrized, and its
+    /// first type parameter should be a constant of type `u128`, which represents the control
+    /// mask. Return a pair of the name of the tag type, as well as the control mask.
+    fn extract_tag_kind_and_mask(&mut self) -> Option<Tag> {
+        precondition!(
+            self.callee_known_name == KnownNames::MiraiAddTag
+                || self.callee_known_name == KnownNames::MiraiHasTag
+                || self.callee_known_name == KnownNames::MiraiDoesNotHaveTag
+        );
+        if let Some(rustc_gen_args) = self.callee_generic_arguments {
+            checked_assume!(rustc_gen_args.len() == 2);
+            if let GenericArgKind::Type(tag_rustc_type) = rustc_gen_args[1].unpack() {
+                if let TyKind::Adt(adt_def, substs_ref) = &tag_rustc_type.kind {
+                    checked_assume!(substs_ref.len() > 0);
+                    if let GenericArgKind::Const(tag_mask_rustc_const) = substs_ref[0].unpack() {
```

Comment:
You can check the ty field of tag_mask_rustc_const to ensure that the constant is actually of type `EnumSet<TagPropagation>`. If not, you can generate a diagnostic here.